internal/pkg/instance/instance_linux_test.go

```diff
@@ -0,0 +1,279 @@
+// LICENSE.md file distributed with the sources of this project regarding your
+// rights to use or distribute this software.
+
+package instance
+
+import (
+	"os"
+	"path/filepath"
+	"testing"
+
+	specs "github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/sylabs/singularity/internal/pkg/test"
+)
+
+const testSubDir = "testing"
+
+func TestProcName(t *testing.T) {
+	test.DropPrivilege(t)
+	defer test.ResetPrivilege(t)
+
+	test := ProcName("test", "test")
+	match := "Singularity instance: test [test]"
+	if test != match {
+		t.Errorf("unexpected match %s != %s", test, match)
+	}
+}
+
+func TestExtractName(t *testing.T) {
+	test.DropPrivilege(t)
+	defer test.ResetPrivilege(t)
+
+	tests := []struct {
+		input  string
+		output string
+	}{
+		{
+			input:  "instance://test",
+			output: "test",
+		},
+		{
+			input:  "instance:/test",
+			output: "instance:/test",
+		},
+		{
+			input:  "instance:///test",
+			output: "/test",
+		},
+	}
+	for _, e := range tests {
+		o := ExtractName(e.input)
+		if o != e.output {
+			t.Errorf("unexpected match, got %s instead of %s", o, e.output)
+		}
+	}
+}
+
+func TestCheckName(t *testing.T) {
+	test.DropPrivilege(t)
+	defer test.ResetPrivilege(t)
+
+	tests := []struct {
+		desc          string
+		name          string
+		expectFailure bool
+	}{
+		{
+			desc:          "with valid name",
+			name:          "test",
+			expectFailure: false,
+		},
+		{
+			desc:          "with valid name containg number",
+			name:          "test123",
+			expectFailure: false,
+		},
+		{
+			desc:          "with invalid name containing space",
+			name:          "test 123",
+			expectFailure: true,
+		},
+		{
+			desc:          "with valid name containing underscore",
+			name:          "test_123",
+			expectFailure: false,
+		},
+		{
+			desc:          "with valid name containing dot",
+			name:          "test.123",
+			expectFailure: false,
+		},
+		{
+			desc:          "with valid name containing minus",
+			name:          "test-123",
+			expectFailure: false,
+		},
+		{
+			desc:          "with invalid name containing slash",
+			name:          "test/123",
+			expectFailure: true,
+		},
+	}
+	for _, e := range tests {
+		err := CheckName(e.name)
+		if err != nil && !e.expectFailure {
+			t.Errorf("unexpected failure %s: %s", e.desc, err)
+		}
+	}
+}
+
+func TestGetDirUnprivileged(t *testing.T) {
+	test.EnsurePrivilege(t)
+
+	hostname, err := os.Hostname()
+	if err != nil {
+		t.Fatalf("unable to retrieve hostname: %s", err)
+	}
+
+	instancePath := filepath.Join("/root", unprivPath, testSubDir, hostname, "root")
+
+	tests := []struct {
+		name          string
+		path          string
+		expectFailure bool
+	}{
+		{
+			name:          "test",
+			path:          filepath.Join(instancePath, "test"),
+			expectFailure: false,
+		},
+		{
+			name:          "test/123",
+			expectFailure: true,
+		},
+	}
+	for _, e := range tests {
+		path, err := GetDirUnprivileged(e.name, testSubDir)
+		if err != nil && !e.expectFailure {
```

Comment:
there's also the case of:

    if err == nil && e.expectFailure {
        t.Errorf("...")
    }

the complete set could be expressed as:

    switch {
    case err == nil && e.expectFailure:
        t.Errorf(...)
    case err != nil && !e.expectFailure:
        t.Errorf(...)
    case err != nil && e.expectFailure:
        // everything OK
    case path != e.path:
        t.Errorf(...)
     }
```